lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -1892,21 +1891,22 @@ def increment_failed_requests_count(code)
     # ingested by the Stackdriver Logging API.
     def increment_ingested_entries_count(count)
       return unless @ingested_entries_count
-      @ingested_entries_count.increment({}, count)
+      @ingested_entries_count.increment({ grpc: @use_grpc, code: @ok_code },
+                                        count)
     end
 
     # Increment the metric for the number of log entries that were dropped
     # and not ingested by the Stackdriver Logging API.
-    def increment_dropped_entries_count(count)
+    def increment_dropped_entries_count(count, code)
```

Comment:
Can you give some background on this change? It looks like you're adding an RPC error code to this metric, but entries could be dropped for other reasons.

Also the Kubernetes Monitoring team may be using this metric. Have you checked with them that the new metric field won't break anything in their existing collection and queries?

Reply:
It will, but only when the new fluentd is rolled out, and that's when we're going to fix stuff

Reply:
@jkohen - The reason we are adding an RPC error code is because we drop / retry entries for various reasons and we want to track the reason.

```
            when \
                # HTTP status 400 (Bad Request).
                GRPC::Core::StatusCodes::INVALID_ARGUMENT,
                # HTTP status 401 (Unauthorized).
                # These are usually solved via a `gcloud auth` call, or by
                # modifying the permissions on the Google Cloud project.
                GRPC::Core::StatusCodes::UNAUTHENTICATED,
                # HTTP status 403 (Forbidden).
                GRPC::Core::StatusCodes::PERMISSION_DENIED,
                # HTTP status 404 (Not Found).
                GRPC::Core::StatusCodes::NOT_FOUND,
                # HTTP status 409 (Conflict).
                GRPC::Core::StatusCodes::ABORTED,
                # HTTP status 412 (Precondition Failed).
                GRPC::Core::StatusCodes::FAILED_PRECONDITION,
                # HTTP status 429 (Too Many Requests).
                GRPC::Core::StatusCodes::RESOURCE_EXHAUSTED,
                # HTTP status 499 (Client Closed Request).
                GRPC::Core::StatusCodes::CANCELLED,
                # the remaining http codes in both 4xx and 5xx category.
                # It's debatable whether to retry or drop these log entries.
                # This decision is made to avoid retrying forever due to client
                # errors.
                GRPC::Core::StatusCodes::UNKNOWN
              increment_dropped_entries_count(entries_count, error.code)
```

These only covers dropped entries due to RPC errors though. On the other hand, we also drop log entries for malformed record/tag.
```
        unless record.is_a?(Hash)
          @log.warn 'Dropping log entries with malformed record: ' \
                    "'#{record.inspect}'. " \
                    'A log record should be in JSON format.'
          next
        end
        sanitized_tag = sanitize_tag(tag)
        if sanitized_tag.nil?
          @log.warn "Dropping log entries with invalid tag: '#{tag.inspect}'." \
                    ' A tag should be a string with utf8 characters.'
          next
        end
```
I've just created a ticket to add counter tracking for those as well.

And yes, this will affect Kubernetes metrics. I've given Mik a heads up of the coming changes. Will update when the change is actually released.